sklearn/tree/tests/test_tree.py

```diff
@@ -3047,3 +3048,16 @@ def test_missing_values_and_constant_toy():
     assert_array_equal(tree.predict(X), y)
     # with just one split (-> three nodes: the root + 2 leaves)
     assert tree.tree_.node_count == 3
+
+
+def test_missing_values_and_poisson_toy():
+    # Test that poisson criterion produces non-negative impurities
+    # when missing values were present.
+    # Non regression test for https://github.com/scikit-learn/scikit-learn/issues/32870
+
+    X = np.array([np.nan, 1, 2, 3, np.nan]).reshape(-1, 1)
+    y = [0.49, 0.5, 0.7, 1.5, 0.8]
+
+    tree = DecisionTreeRegressor(criterion="poisson")
+    tree.fit(X, y)
+    assert (tree.tree_.impurity >= -1e10).all()
```

Comment:
Can we just fold this test into the above `test_regression_tree_missing_values_toy`? In the parametrization, you can add a tolerance, and use the docstring/inline comments to help elaborate what's going on.